### PR TITLE
"dnu restore" doesn't cache invalid contents

### PR DESCRIPTION
**Pull request by ChengTian**
parent https://github.com/aspnet/dnx/issues/1746

Use the approach suggested by @davidfowl 

Also did some cleanup, replacing `string.Format()` with string interpolation.

Also fixing https://github.com/aspnet/dnx/issues/1804 by adding just one statement.

### CODE REVIEW

File: src/Microsoft.Framework.PackageManager/Restore/NuGet/NuGetv2Feed.cs [hunk 1]
```diff
@@ -239,17 +243,59 @@ public PackageInfo BuildModel(string id, XElement element)
                 {
                     if (retry == 2)
                     {
-                        _reports.Error.WriteLine(string.Format("Error: DownloadPackageAsync: {1}\r\n  {0}", ex.Message, package.ContentUri.Red().Bold()));
+                        _reports.Error.WriteLine(
+                            $"Error: DownloadPackageAsync: {package.ContentUri}{Environment.NewLine}  {ex.Message}".Red().Bold());
+                        throw;
```
Comment (ChengTian) [inline src/Microsoft.Framework.PackageManager/Restore/NuGet/NuGetv2Feed.cs]: Adding re-throw here can fix https://github.com/aspnet/dnx/issues/1804

**Comment by ChengTian** [comment]
@davidfowl , using file stream to avoid memory explosion now.

File: src/Microsoft.Framework.PackageManager/Restore/NuGet/HttpSource.cs [hunk 2]
```diff
@@ -117,67 +118,87 @@ internal class HttpSource
...
-            response.EnsureSuccessStatusCode();
-
-            var newFile = result.CacheFileName + "-new";
-
-            // Zero value of TTL means we always download the latest package
-            // So we write to a temp file instead of cache
-            if (cacheAgeLimit.Equals(TimeSpan.Zero))
+            // Write the response to a file and use the file stream to avoid memory explosion
+            using (var responseStream = new FileStream(
+                Path.GetTempFileName(),
```
Comment (davidfowl) [inline src/Microsoft.Framework.PackageManager/Restore/NuGet/HttpSource.cs]: Do we make sure this file is always deleted?

Comment (ChengTian) [inline src/Microsoft.Framework.PackageManager/Restore/NuGet/HttpSource.cs]: Yes, see `FileOptions.DeleteOnClose` option below :smile:

**Comment by davidfowl** [comment]
:shipit:

**Pull request merged**